File: spine-cocos2dx/example/Classes/ExampleLayer.cpp


#include "ExampleLayer.h"
#include <iostream>
#include <fstream>
#include <string.h>

using namespace cocos2d;
using namespace spine;
using namespace std;

CCScene* ExampleLayer::scene () {
	CCScene *scene = CCScene::create();
	scene->addChild(ExampleLayer::create());
	return scene;
}

bool ExampleLayer::init () {
	if (!CCLayer::init()) return false;
	
	skeletonNode = CCSkeletonAnimation::createWithFile("spineboy.json", "spineboy.atlas");
	skeletonNode->setMix("walk", "jump", 0.2f);
	skeletonNode->setMix("jump", "walk", 0.4f);
	
	skeletonNode->setAnimationListener(this, animationStateEvent_selector(ExampleLayer::animationStateEvent));
	skeletonNode->setAnimation(0, "walk", false);
	skeletonNode->addAnimation(0, "jump", false);
	skeletonNode->addAnimation(0, "walk", true);
	skeletonNode->addAnimation(0, "jump", true, 4);
	skeletonNode->addAnimation(1, "drawOrder", true);

	// skeletonNode->timeScale = 0.3f;
	skeletonNode->debugBones = true;

	skeletonNode->runAction(CCRepeatForever::create(CCSequence::create(CCFadeOut::create(1),
		CCFadeIn::create(1),
		CCDelayTime::create(5),
		NULL)));

	CCSize windowSize = CCDirector::sharedDirector()->getWinSize();
	skeletonNode->setPosition(ccp(windowSize.width / 2, 20));
	addChild(skeletonNode);

	scheduleUpdate();

	return true;
}

void ExampleLayer::update (float deltaTime) {
	// Test releasing memory.
	// if (entry->time > 0.1) CCDirector::sharedDirector()->replaceScene(ExampleLayer::scene());
}

void ExampleLayer::animationStateEvent (CCSkeletonAnimation* node, int trackIndex, EventType type, Event* event, int loopCount) {
	TrackEntry* entry = AnimationState_getCurrent(node->state, trackIndex);
	const char* animationName = (entry && entry->animation) ? entry->animation->name : 0;

	switch (type) {
	case ANIMATION_START:
		CCLog("%d start: %s", trackIndex, animationName);
		break;
	case ANIMATION_END:
		CCLog("%d end: %s", trackIndex, animationName);
		break;
	case ANIMATION_COMPLETE:
		CCLog("%d complete: %s, %d", trackIndex, animationName, loopCount);
		break;
	case ANIMATION_EVENT:
		CCLog("%d event: %s, %s: %d, %f, %s", trackIndex, animationName, event->data->name, event->intValue, event->floatValue, event->stringValue);
		break;
	}
	fflush(stdout);
}

File: spine-cocos2d-iphone/example/ExampleLayer.m


#import "ExampleLayer.h"

@implementation ExampleLayer

+ (CCScene*) scene {
	CCScene *scene = [CCScene node];
	[scene addChild:[ExampleLayer node]];
	return scene;
}

-(id) init {
	self = [super init];
	if (!self) return nil;

	animationNode = [CCSkeletonAnimation skeletonWithFile:@"spineboy.json" atlasFile:@"spineboy.atlas" scale:1];
	[animationNode setMixFrom:@"walk" to:@"jump" duration:0.2f];
	[animationNode setMixFrom:@"jump" to:@"walk" duration:0.4f];
	[animationNode setDelegate:self];
	[animationNode setAnimationForTrack:0 name:@"walk" loop:NO];
	[animationNode addAnimationForTrack:0 name:@"jump" loop:NO afterDelay:0];
	[animationNode addAnimationForTrack:0 name:@"walk" loop:YES afterDelay:0];
	[animationNode addAnimationForTrack:0 name:@"jump" loop:YES afterDelay:4];
	[animationNode setAnimationForTrack:1 name:@"drawOrder" loop:YES];
	animationNode.timeScale = 0.3f;
	animationNode.debugBones = true;

	CGSize windowSize = [[CCDirector sharedDirector] winSize];
	[animationNode setPosition:ccp(windowSize.width / 2, 20)];
	[self addChild:animationNode];

#if __CC_PLATFORM_MAC
	[self setMouseEnabled:YES];
#endif

	return self;
}

- (void) animationDidStart:(CCSkeletonAnimation*)animation track:(int)trackIndex {
	CCLOG(@"%d start: %s", trackIndex, AnimationState_getCurrent(animation.state, trackIndex)->animation->name);
}

- (void) animationWillEnd:(CCSkeletonAnimation*)animation track:(int)trackIndex {
	CCLOG(@"%d end: %s", trackIndex, AnimationState_getCurrent(animation.state, trackIndex)->animation->name);
}

- (void) animationDidTriggerEvent:(CCSkeletonAnimation*)animation track:(int)trackIndex event:(Event*)event {
	CCLOG(@"%d event: %s, %s: %d, %f, %s", trackIndex, AnimationState_getCurrent(animation.state, trackIndex)->animation->name,
			event->data->name, event->intValue, event->floatValue, event->stringValue);
}

- (void) animationDidComplete:(CCSkeletonAnimation*)animation track:(int)trackIndex loopCount:(int)loopCount {
	CCLOG(@"%d complete: %s, %d", trackIndex, AnimationState_getCurrent(animation.state, trackIndex)->animation->name, loopCount);
}

#if __CC_PLATFORM_MAC
- (BOOL) ccMouseDown:(NSEvent*)event {
	CCDirector* director = [CCDirector sharedDirector];
	NSPoint location =  [director convertEventToGL:event];
	location.x -= [[director runningScene]position].x;
	location.y -= [[director runningScene]position].y;
	location.x -= animationNode.position.x;
	location.y -= animationNode.position.y;
	if (CGRectContainsPoint(animationNode.boundingBox, location)) NSLog(@"Clicked!");
	return YES;
}
#endif

@end